### PR TITLE
[ENH] Add the ability to check the validity of an MAG

### PR DESCRIPTION
**Pull request by aryan26roy**
Closes #90 

Changes proposed in this pull request:

- Add a function and related helper functions to check the validity of an MAG.

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by aryan26roy** [comment]
@adam2392 any idea why the DCO is complaining? I have signed off my commits.

**Comment by adam2392** [comment]
> @adam2392 any idea why the DCO is complaining? I have signed off my commits.

I wouldn't worry about it. You can try doing the rebasing instructions to fix it if you want: https://github.com/py-why/pywhy-graphs/pull/91/checks?check_run_id=15461061662

**Comment by aryan26roy** [comment]
@adam2392 In finding directed cycles, does the direction of the directed edge matter? 
For eg., is this a directed cycle?  a -> b -> c -> d <- a

**Comment by adam2392** [comment]
> @adam2392 In finding directed cycles, does the direction of the directed edge matter?
> For eg., is this a directed cycle? a -> b -> c -> d <- a

Sorry for the delay.

In causality, typically cycles are directed cycles. In directed cycles, by definition the direction matters. `a -> b -> c -> d <- a` is not a cycle, but `a -> b -> c -> d -> a` is.

I would review the [Zhang paper](https://www.sciencedirect.com/science/article/pii/S0004370208001008) if any questions like this arise in case I can't respond in a timely manner.

**Comment by aryan26roy** [comment]
@adam2392 thanks for the clarification.

**Comment by aryan26roy** [comment]
@adam2392 What should I have in the sets L and S when checking for an inducing path?

**Comment by adam2392** [comment]
> @adam2392 What should I have in the sets L and S when checking for an inducing path?

You can have them be passed in by user and default to empty sets.

**Comment by aryan26roy** [comment]
@adam2392 I think the function is complete. Now I should add unit tests. Do you know how I can get examples of valid and invalid MAGs?

**Comment by adam2392** [comment]
> @adam2392 I think the function is complete. Now I should add unit tests. Do you know how I can get examples of valid and invalid MAGs?

I would just construct them by hand following the definition, or look at the papers on causal discovery (if you see them in a paper, cite where you got it from and which figure).

File: pywhy_graphs/algorithms/generic.py [hunk 1]
```diff
@@ -604,3 +605,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
                 break
 
     return (path_exists, path)
+
+
+def _find_adc(G):
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:610]: This can be a public function and tested separately to help you ensure that this works as intended.

Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:610]: Sorry for the late response, you're right. I'll do just that.

File: pywhy_graphs/algorithms/tests/test_generic.py [hunk 2]
```diff
@@ -228,3 +228,102 @@ def test_is_collider():
...
+    admg.add_edge("H", "Z", admg.directed_edge_name)
+    admg.add_edge("K", "H", admg.directed_edge_name)
+    admg.add_edge("Y", "H", admg.directed_edge_name)
+    admg.add_edge("K", "J", admg.bidirected_edge_name)
+
+    assert pywhy_graphs.find_adc(admg)
+
+
+def test_valid_mag():
+    # K -> H -> Z -> X -> Y -> J <- K
```
Comment (adam2392) [inline pywhy_graphs/algorithms/tests/test_generic.py:285]: Can you add reasonings in the in-line comments for why these aren't valid MAGs?

File: pywhy_graphs/algorithms/tests/test_generic.py [hunk 3]
```diff
@@ -228,3 +228,102 @@ def test_is_collider():
     S = {"A"}
 
     assert pywhy_graphs.inducing_path(admg, "Z", "Y", L, S)[0]
+
+
+def test_find_adc():
+    # K -> H -> Z -> X -> Y -> J <- K
```
Comment (adam2392) [inline pywhy_graphs/algorithms/tests/test_generic.py:234]: Can you add reasonings in the in-line comments for why these are/aren't valid ADC?

Comment (aryan26roy) [reply pywhy_graphs/algorithms/tests/test_generic.py:234]: Sure!

File: pywhy_graphs/algorithms/tests/test_generic.py (hunk 2 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/tests/test_generic.py:285]: Alright.

**Comment by aryan26roy** [comment]
@adam2392 Is this good enough?

**Comment by adam2392** [comment]
Can you add a CHANGELOG entry to fix the CI?

And add these functions to `api.rst`?

File: pywhy_graphs/algorithms/generic.py [hunk 4]
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
...
+
+    biedges = G.bidirected_edges
+
+    for elem in G.nodes:
+        ancestors = nx.ancestors(G.sub_directed_graph(), elem)
+        descendants = nx.descendants(G.sub_directed_graph(), elem)
+        for elem in biedges:
+            if (elem[0] in ancestors and elem[1] in descendants) or (
+                elem[1] in ancestors and elem[0] in descendants
+            ):  # there is an undirected edge from one of the ancestors to a descendant
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:636]: undirected or bidirected?

File: pywhy_graphs/algorithms/generic.py [hunk 5]
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
                 break
 
     return (path_exists, path)
+
+
+def find_adc(G):
+
+    """Finds an Almost Directed Cycles in a mixed edge graph.
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:613]: If you name it `find_*`, it should return the actual path found: E.g. see similar function in networkx: https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.cycles.find_cycle.html

Otw, you can name it `has_almost_directed_cycle`, which makes more sense. Or you can return the actual path that is the "almost directed cycle".

File: pywhy_graphs/algorithms/generic.py [hunk 6]
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
...
+
+    Returns
+    -------
+    is_valid : bool
+        A boolean indicating whether the provided graph is a valid MAG or not.
+
+    """
+
+    if L is None:
+        L = set()
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:658]: This line is not tested.

File: pywhy_graphs/algorithms/generic.py [hunk 7]
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
...
+
+    if S is None:
+        S = set()
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:661]: This line is not tested.

File: pywhy_graphs/algorithms/generic.py [hunk 8]
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
...
+    for source in all_nodes:
+        nb = set(G.neighbors(source))
+        cur_set = all_nodes - nb
+        cur_set.remove(source)
+        for dest in cur_set:
+            out = inducing_path(G, source, dest, L, S)
+            if out[0] is True:
+                return False
+
+    return is_valid
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:690]: just `return True` since `is_valid` is not used?

File: pywhy_graphs/algorithms/generic.py [hunk 9]
```diff
@@ -604,3 +606,85 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
...
+        for elem in biedges:
+            if (elem[0] in ancestors and elem[1] in descendants) or (
+                elem[1] in ancestors and elem[0] in descendants
+            ):  # there is an undirected edge from one of the ancestors to a descendant
+                return not adc_present
+
+    return adc_present
+
+
+def valid_mag(G: ADMG, L: set = None, S: set = None):
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:642]: An additional check perhaps to add is that there is only one edge between any two nodes.

I.e. in a MAG, this is not possible: `G = (X -> Y; X <--> Y)`

File: pywhy_graphs/algorithms/generic.py (hunk 4 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:636]: Ah error in the comments, should be bidirected.

File: pywhy_graphs/algorithms/generic.py (hunk 5 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:613]: Got it.

File: pywhy_graphs/algorithms/generic.py (hunk 8 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:690]: alright

File: pywhy_graphs/algorithms/generic.py (hunk 9 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:642]: I will also add a check for undirected edges (Since ADMG can have those).

**Comment by aryan26roy** [comment]
@adam2392 I updated the CHANGELOG for v0.2, is that fine?

File: pywhy_graphs/algorithms/generic.py [hunk 10]
```diff
@@ -604,3 +606,93 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
...
+                elem[1] in ancestors and elem[0] in descendants
+            ):  # there is a bidirected edge from one of the ancestors to a descendant
+                return not adc_present
+
+    return adc_present
+
+
+def valid_mag(G: ADMG, L: set = None, S: set = None):
+    """Checks if the provided graph is a valid MAG.
+
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:643]: The docstring should describe what a "valid MAG" is.

File: pywhy_graphs/algorithms/generic.py [hunk 11]
```diff
@@ -604,3 +606,93 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
                 break
 
     return (path_exists, path)
+
+
+def has_adc(G):
+    """Finds if a graph has an Almost Directed Cycle.
+
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:613]: The docstring should describe what an "almost directed cycle" is

File: pywhy_graphs/algorithms/generic.py [hunk 12]
```diff
@@ -604,3 +606,93 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
                 break
 
     return (path_exists, path)
+
+
+def has_adc(G):
+    """Finds if a graph has an Almost Directed Cycle.
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:612]:
```suggestion
    """Check if a graph has an almost directed cycle.
```

**Review by adam2392 — COMMENTED** [review]
Few nits and then it LGTM. This will now help you finish https://github.com/py-why/pywhy-graphs/issues/85

Which will finally enable you to finish the PR for checking validity of a PAG perhaps(?).

We can always revisit this code if there is a bug you find

**Comment by aryan26roy** [comment]
@adam2392 I added the definitions, can you take a look?

File: pywhy_graphs/algorithms/generic.py [hunk 13]
```diff
@@ -604,3 +606,100 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
                 break
 
     return (path_exists, path)
+
+
+def has_adc(G):
+    """Check if a graph has an almost directed cycle.
+
+    An almost directed cycle is a is a directed cycle containing
+    one bidirected edge.
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:615]:
```suggestion
    one bidirected edge. For example, ``A -> B -> C <-> A`` is an adc.
```

**Review by adam2392 — APPROVED** [review]
LGTM after final few comments

File: pywhy_graphs/algorithms/generic.py [hunk 14]
```diff
@@ -604,3 +606,100 @@ def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
...
+
+    return adc_present
+
+
+def valid_mag(G: ADMG, L: set = None, S: set = None):
+    """Checks if the provided graph is a valid MAG.
+
+    A valid MAG is a mixed edge graph that only has directed and bi-directed
+    edges, no directed or almost directed cycles and no inducing paths between
+    any two non-adjacent pair of nodes.
```
[the remark this answers is not in the document]
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:649]: Sure

**Comment by aryan26roy** [comment]
@adam2392 I rebased the branch and removed the quotes from my email address but the DCO won't stop complaining. Is there a way to merge this without DCO approval?

**Review by adam2392 — APPROVED** [review]

**Comment by adam2392** [comment]
Thanks @aryan26roy !

**Pull request merged**